### PR TITLE
fix: prevent leak of files to /dev/shm in corner cases (#17658)

### LINKED ISSUE #17658: 'util/git/creds.go' will leak temporary files to '/dev/shm' if proxy URL cannot be parsed, or if write fails
Checklist:

* [X] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [X] I've included steps to reproduce the bug.
* [X] I've pasted the output of `argocd version`.

**Describe the bug**

In `util/git/creds.go`, within `(c SSHCreds) Environ()`, there are a couple of corner cases where the temporary file that is created is not being cleaned up:
- If the socks5 proxy settings URL cannot be parsed by `url.Parse`, the function will return without deleting the temporary file.
- If a call to `file.WriteString` fails, the temporary file will not be cleaned up (but, admittedly, this would be very rare)

We've had customer reports of `/dev/shm` running out of space due to temporary files left hanging around, and I discovered the above issue while auditing Argo CD's use of `/dev/shm`.

I've submitted a PR which fixes both issues. The solution is the same one used by the other `Environ()` functions in `creds.go`

**Version**

`master` branch

### PR DESCRIPTION
**Pull request by jgwest**
Fixes #17658

See parent issue for description of the problem. TL;DR: we're not cleaning up temp files from /dev/shm in a couple corner cases.

The solution is simple: 
- ensure that `Close` is always called on the `io.Closer`, which causes the temporary file to be deleted. 
- we follow the same pattern used by the other [`Environ()` functions in the same file](https://github.com/argoproj/argo-cd/blob/e26f4fbdc12a98ea3801c1128141dde211afa2dc/util/git/creds.go#L199)

Checklist:

* [X] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [X] The title of the PR states what changed and the related issues number (used for the release note).
* [X] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [X] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [X] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [X] Does this PR require documentation updates?
* [X] I've updated documentation as required by this PR.
* [X] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [X] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [X] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [X] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [X] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [X] Optional. My organization is added to USERS.md.
* [X] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

### CODE REVIEW

File: docs/operator-manual/server-commands/argocd-repo-server.md [hunk 1]
```diff
@@ -21,6 +21,7 @@ argocd-repo-server [flags]
       --disable-helm-manifest-max-extracted-size       Disable maximum size of helm manifest archives when extracted
       --disable-tls                                    Disable TLS on the gRPC endpoint
       --helm-manifest-max-extracted-size string        Maximum size of helm manifest archives when extracted (default "1G")
+      --helm-registry-max-index-size string            Maximum size of registry index file (default "1G")
```
Comment (jgwest) [inline docs/operator-manual/server-commands/argocd-repo-server.md:24]: Fix CI failure from a previous PR

**Review by jannfis — COMMENTED** [review]
Can we come up with a test for that to prevent regressions in the future?

**Comment by jgwest** [comment]
@jannfis Anything specific you had in mind? Only thing I can think of is mocking the specific corner cases that caused the leak in the first place, which, given that they are very specific corner cases, are unlikely to reoccur, and  mocking them would be a lot of work for minimal gain.

(Much much more likely is that one of the calling functions is going to forget to close the `io.Closer` which is returned by this function. I'd expect the only way to check for would be some kind of fancy static analysis.)

 WDYT?

**Review by jannfis — APPROVED** [review]
LGTM

**Comment by jannfis** [comment]
@jgwest I think that's good enough, thank you!

**Pull request merged**